File: src/utils/tokenStore.ts
import { invoke } from "@tauri-apps/api/core";

// Where the two Google OAuth tokens now live, and why (see AUDIT.md S1):
//
// - access_token: short-lived (~1h), kept in a plain in-memory module
//   variable. Never persisted anywhere -- not localStorage, not the OS
//   keychain. There's nothing gained by persisting it (a fresh
//   refresh-token exchange re-derives it in one round-trip on every app
//   start) and every extra place a live token is written at rest is one
//   more thing an attacker could read.
// - refresh_token: long-lived (valid indefinitely until revoked or unused
//   for ~6 months), stored via the Rust `store_token`/`get_token`/
//   `clear_token` commands, which delegate to the OS's own credential
//   store (Windows Credential Manager / macOS Keychain / Linux Secret
//   Service) through the `keyring` crate. This replaces the previous
//   plaintext `localStorage` storage for both tokens.

let accessToken: string | null = null;
let tokenIssuedAt = 0;

export function getAccessToken(): string | null {
  return accessToken;
}

/** Epoch ms the current access token was issued/refreshed at, or 0 if none. */
export function getTokenIssuedAt(): number {
  return tokenIssuedAt;
}

export function setAccessToken(token: string | null, issuedAt: number = Date.now()): void {
  accessToken = token;
  tokenIssuedAt = token ? issuedAt : 0;
}

const REFRESH_TOKEN_ACCOUNT = "refresh_token";

export async function storeRefreshToken(token: string): Promise<void> {
  await invoke("store_token", { account: REFRESH_TOKEN_ACCOUNT, value: token });
}

export async function getRefreshToken(): Promise<string | null> {
  try {
    return await invoke<string | null>("get_token", { account: REFRESH_TOKEN_ACCOUNT });
  } catch (e) {
    // No OS keychain available (e.g. headless Linux with no Secret Service
    // daemon running) or some other platform error -- treat as "not logged
    // in" rather than throwing, so callers fall back to the login screen
    // instead of crashing.
    console.warn("[tokenStore] get_token failed (treating as no stored refresh token)", e);
    return null;
  }
}

export async function clearRefreshToken(): Promise<void> {
  try {
    await invoke("clear_token", { account: REFRESH_TOKEN_ACCOUNT });
  } catch (e) {
    console.warn("[tokenStore] clear_token failed (non-blocking)", e);
  }
}

File: src-tauri/src/commands/token_store.rs
//! Secure OS-keychain-backed storage for long-lived OAuth credentials.
//!
//! Replaces storing the Google refresh token in the frontend's `localStorage`
//! (plaintext, unencrypted-at-rest, readable by any process running as the
//! same OS user) with the `keyring` crate, which delegates to the OS's own
//! credential store: Windows Credential Manager, macOS Keychain, or the
//! Linux Secret Service (GNOME Keyring / KWallet via D-Bus).
//!
//! The access token deliberately does NOT go through here -- it is short-
//! lived (~1h) and kept in-memory only on the frontend side (see
//! `src/utils/tokenStore.ts`), so there's no round-trip IPC cost on every
//! check and nothing meaningful is lost by not persisting it at all across a
//! restart (a fresh refresh-token exchange re-establishes it immediately).
use tauri::command;

use crate::AppError;

/// Keychain "service" namespace. Paired with an `account` string
/// (currently only "refresh_token") to form the OS credential's identity.
const SERVICE_NAME: &str = "drplay";

#[command]
pub async fn store_token(account: String, value: String) -> Result<(), AppError> {
    tauri::async_runtime::spawn_blocking(move || {
        let entry = keyring::Entry::new(SERVICE_NAME, &account).map_err(|e| AppError::Keychain(e.to_string()))?;
        entry.set_password(&value).map_err(|e| AppError::Keychain(e.to_string()))
    })
    .await
    .map_err(|e| AppError::TaskPanicked(format!("Task panicked: {e}")))?
}

#[command]
pub async fn get_token(account: String) -> Result<Option<String>, AppError> {
    tauri::async_runtime::spawn_blocking(move || {
        let entry = keyring::Entry::new(SERVICE_NAME, &account).map_err(|e| AppError::Keychain(e.to_string()))?;
        match entry.get_password() {
            Ok(pw) => Ok(Some(pw)),
            Err(keyring::Error::NoEntry) => Ok(None),
            Err(e) => Err(AppError::Keychain(e.to_string())),
        }
    })
    .await
    .map_err(|e| AppError::TaskPanicked(format!("Task panicked: {e}")))?
}

#[command]
pub async fn clear_token(account: String) -> Result<(), AppError> {
    tauri::async_runtime::spawn_blocking(move || {
        let entry = keyring::Entry::new(SERVICE_NAME, &account).map_err(|e| AppError::Keychain(e.to_string()))?;
        match entry.delete_credential() {
            Ok(()) => Ok(()),
            // Already absent -- logout/clear should be idempotent, not an error.
            Err(keyring::Error::NoEntry) => Ok(()),
            Err(e) => Err(AppError::Keychain(e.to_string())),
        }
    })
    .await
    .map_err(|e| AppError::TaskPanicked(format!("Task panicked: {e}")))?
}
